mapviz_plugins/src/object_plugin.cpp

```diff
@@ -0,0 +1,467 @@
+    {
+      tf::Vector3 pt(point.x, point.y, point.z);
+      pt = data.local_transform*pt;
+      tf::Vector3 transformed = transform *pt;
+      data.polygon.push_back({pt, transformed});
+    }
+    if (data.polygon.size())
+    {
+      data.polygon.push_back(data.polygon.front());
+    }
+    objects_.push_back(data);
+  }
+
+  void ObjectPlugin::handleTrack(const marti_nav_msgs::TrackedObject &obj)
+  {
+    if (!has_message_)
+    {
+      initialized_ = true;
+      has_message_ = true;
+    }
+
+    // Since orientation was not implemented, many markers publish
+    // invalid all-zero orientations, so we need to check for this
+    // and provide a default identity transform.
+    tf::Quaternion orientation(0.0, 0.0, 0.0, 1.0);
+    if (obj.pose.pose.orientation.x ||
+        obj.pose.pose.orientation.y ||
+        obj.pose.pose.orientation.z ||
+        obj.pose.pose.orientation.w)
+    {
+      orientation = tf::Quaternion(obj.pose.pose.orientation.x,
+                                   obj.pose.pose.orientation.y,
+                                   obj.pose.pose.orientation.z,
+                                   obj.pose.pose.orientation.w);
+    }
+
+    ObjectData data;
+    data.local_transform = swri_transform_util::Transform(
+      tf::Transform(
+        orientation,
+        tf::Vector3(obj.pose.pose.position.x,
+                    obj.pose.pose.position.y,
+                    obj.pose.pose.position.z)));
+    data.stamp = obj.header.stamp;
+    data.source_frame = obj.header.frame_id;
+    data.id = std::to_string(obj.id);
+    data.transformed = false;
+
+    swri_transform_util::Transform transform;
+    if (GetTransform(data.source_frame, data.stamp, transform))
+    {
+      data.transformed = true;
+    }
+
+    data.polygon.reserve(obj.polygon.size()+1);
+    for (auto& point: obj.polygon)
+    {
+      tf::Vector3 pt(point.x, point.y, point.z);
+      pt = data.local_transform*pt;
+      tf::Vector3 transformed = transform *pt;
+      data.polygon.push_back({pt, transformed});
+    }
+    if (data.polygon.size())
+    {
+      data.polygon.push_back(data.polygon.front());
+    }
+    objects_.push_back(data);
+  }
+
+  void ObjectPlugin::PrintError(const std::string& message)
+  {
+    PrintErrorHelper(ui_.status, message);
+  }
+
+  void ObjectPlugin::PrintInfo(const std::string& message)
+  {
+    PrintInfoHelper(ui_.status, message);
+  }
+
+  void ObjectPlugin::PrintWarning(const std::string& message)
+  {
+    PrintWarningHelper(ui_.status, message);
+  }
+
+  QWidget* ObjectPlugin::GetConfigWidget(QWidget* parent)
+  {
+    config_widget_->setParent(parent);
+
+    return config_widget_;
+  }
+
+  bool ObjectPlugin::Initialize(QGLWidget* canvas)
+  {
+    canvas_ = canvas;
+    SetColor(ui_.color->color());
+
+    return true;
+  }
+
+  void ObjectPlugin::Draw(double x, double y, double scale)
+  {
+    auto objIter = objects_.begin();
+    while (objIter != objects_.end())
```

Comment:
Using a while loop and manually incrementing the iterator seems like a weird way to do this, any reason to not use a for-each loop?